src/main/java/io/cdap/plugin/gcp/common/GCPUtils.java

```diff
@@ -49,11 +64,55 @@ public static ServiceAccountCredentials loadServiceAccountCredentials(String pat
     }
   }
 
+  public static ServiceAccountCredentials loadServiceAccountCredentials(String content,
+                                                                        boolean isServiceAccountFilePath)
+    throws IOException {
+    if (isServiceAccountFilePath) {
+      return loadServiceAccountCredentials(content);
+    }
+    InputStream jsonInputStream = new ByteArrayInputStream(content.getBytes());
+    return ServiceAccountCredentials.fromStream(jsonInputStream);
+  }
+
+  public static String extractPrivateKey(ServiceAccountCredentials credentials) {
+    return String.format(PRIVATE_KEY_WRAP,
+                         Base64.getEncoder().encodeToString(credentials.getPrivateKey().getEncoded()));
+  }
+
+  public static Map<String, String> generateAuthProperties(String serviceAccount,
+                                                           boolean isServiceAccountFilePath,
+                                                           String... keyPrefix) throws IOException {
+    Map<String, String> properties = new HashMap<>();
+    String privateKeyData = null;
+    properties.put(SERVICE_ACCOUNT_TYPE, isServiceAccountFilePath ? SERVICE_ACCOUNT_TYPE_FILE_PATH
+      : SERVICE_ACCOUNT_TYPE_JSON);
+
+    for (String prefix : keyPrefix) {
+      if (isServiceAccountFilePath) {
+        properties.put(String.format("%s.%s", prefix, CLOUD_JSON_KEYFILE_SUFFIX), serviceAccount);
+        continue;
+      }
+      ServiceAccountCredentials credentials = loadServiceAccountCredentials(serviceAccount, false);
+
+      properties.put(String.format("%s.%s", prefix, CLOUD_ACCOUNT_EMAIL_SUFFIX), credentials.getClientEmail());
+      properties.put(String.format("%s.%s", prefix, CLOUD_ACCOUNT_PRIVATE_KEY_ID_SUFFIX),
+                     credentials.getPrivateKeyId());
+      if (privateKeyData == null) {
```

Comment:
Isnt private key always null? Why do we have this check?

Reply:
Is kind of optimization. Check is inside loop, and private key will be always same, in first pass we do extract and save in  local variable, so in the second pass key will be already extracted. We can remove the check ?